migrate to immutable.js

jlongster opened the pull request:
I haven't tested this extensively, but the basic features work. I'd like to merge the migration incrementally, even if it's likely that there are bugs.

jlongster commented on js/components/Editor.js:
```diff
@@ -59,7 +59,9 @@ const Editor = React.createClass({
   },
 
   componentDidUpdate() {
-    this.editor.setValue(this.props.sourceText);
```
This and `componentWillReceiveProp` both need a lot of work. I'd say we mostly ignore this code for now until we focus on the editor. For some reason this code was giving me problems and it worked without it.

jasonLaster replied:
agreed. figured that was a placeholder at the time

jasonLaster commented on js/components/Breakpoints.js:
```diff
@@ -9,7 +9,7 @@ require("./Breakpoints.css");
 
 const Breakpoints = React.createClass({
   propTypes: {
-    breakpoints: PropTypes.array,
```
it'd be nice to have better annotations at somepoint

jlongster replied:
Just found this: https://www.npmjs.com/package/react-immutable-proptypes

jasonLaster commented on js/components/Tabs.js:
```diff
@@ -29,13 +27,17 @@ function Tabs({ tabs, selectTab, loadSources, newSource }) {
+          { className: "tab-title" },
+          tab.get("title"),
+          dom.div({ className: "tab-url" }, tab.get("url"))
+        )
```
nice. looks much nicer.

jlongster replied:
I just realized I think I messed up the structure here. `tab-url` shouldn't be inside of `tab-title`. Doh! You're structure before was pretty good but the main thing was moving the props down. Otherwise my JS formatter wants to bump the rest of the args over to line up with it...

Anyway, I'll make a new commit that fixes the structure

jlongster commented:
I rebased this so it can be merged. I added the react-immutable-proptypes package and it's really nice because it makes it clear what is immutable.js and what isn't. Feels good.

jlongster commented:
I will squash these commits if you r+

jasonLaster commented:
r+. squash and :ship:

jlongster commented:
Squashed and :ship:'ed!
